### PR TITLE
Split single package into separate `cli/` and `stubs/` packages

### PR DESCRIPTION
**Pull request by cdaly333**
This PR splits up command line UI layer and the core stubs/radio implementation into their own separate packages.  This achieves a couple main things:

- **Abstracts stubs from UI to enable the modularize UI layer.  This makes way for the chrome implementation.**
- Removed `yargs` dependency from stub package. 

The CLI package imports the stub package and implements a UI layer on top of it.

**Core changes:**
- Added `stubs/` and `cli` directories with new node packages.
- Added `RadioController` interface. (to allow abstract radio controller implementations - namely the chrome version)
- Renamed `RadioController` class to `NodeRadioController`.
- Remove `yargs` dependency from core stub package.
- Moved `UDPScanConfig`, `RemoteInfo` and `UDPScanResults` to `dataflow.ts`.
- Removed redundant dependencies from `package.json`s 

**Minor tweaks:**
- Minor `README.md` updates.
- Remove some unneeded `console.log` calls.

**Issues:**
- Progresses #11

### CODE REVIEW

File: cli/src/command-processor.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import yargs from 'yargs/yargs';
 import * as readline from 'readline';
 import {CommandMessage, READY_FOR_MESSAGE, CHECK_READY} from './commands';
 import {ExecuteMessage, IdentifyMessage, ScanMessage} from './commands';
-import {UDPScanConfig} from '../radio/radio-controller';
+import {UDPScanConfig} from 'local-home-testing/build/src/radio';
```
Comment (Fleker) [inline cli/src/command-processor.ts:6]: Do they need to come from `/build` or can the values be pulled directly from the Typescript source?

File: stubs/src/radio/radio-controller.ts [hunk 2]
```diff
@@ -1,63 +1,58 @@
...
 /**
  * A default radio timeout, in milliseconds.
  */
 const RADIO_TIMEOUT = 2500;
-
 /**
- * A class to contain all parameters required to perform
- * a UDP scan.
+ * A helper function to create timeout promises.
+ * @returns  A new Promise that resolves after RADIO_TIMEOUT milliseconds.
```
Comment (Fleker) [inline stubs/src/radio/radio-controller.ts:10]: This doesn't resolve after `RADIO_TIMEOUT` but after a parameter.

File: cli/README.md [hunk 3]
```diff
@@ -8,11 +8,11 @@ TODO(cjdaly) update usage
```
Comment (proppy) [inline cli/README.md:8]: maybe do those in a separate PR?

File: cli/package.json [hunk 4]
```diff
@@ -0,0 +1,58 @@
+{
+  "name": "local-home-testing-cli",
+  "version": "0.0.1",
+  "description": "A command-line interface for interacting with Local Home SDK stubs.",
```
Comment (proppy) [inline cli/package.json:4]: is that actually using the stubs of the new radio interfaces?

File: cli/package.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  "private": true,
```
Comment (proppy) [inline cli/package.json:5]: should that still be private?

File: cli/package.json [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  "author": "Google LLC.",
+  "license": "Apache-2.0",
+  "main": "./build/src/index.js",
+  "scripts": {
+    "compile": "tsc",
+    "test": "ava",
+    "clean": "gts clean",
+    "cli": "node ./build/src/index.js",
```
Comment (proppy) [inline cli/package.json:13]: ts-node?

File: cli/package.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    "lint": "eslint './src/**/*.ts' './test/**/*.ts'"
```
Comment (proppy) [inline cli/package.json:14]: should that be `gts check` ? (i.e: what result of `gts init`)

File: cli/package.json [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  },
+  "devDependencies": {
+    "ava": "^3.10.0",
+    "@ava/typescript": "^1.1.1",
+    "@google/local-home-sdk": "^1.1.0",
+    "@types/node": "^13.11.1",
+    "@types/yargs": "^15.0.5",
+    "@typescript-eslint/eslint-plugin": "^3.7.0",
+    "chalk": "^4.0.0",
+    "eslint": "^7.5.0",
```
Comment (proppy) [inline cli/package.json:28]: shouldn't that come for free with gts?

File: cli/package.json [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    "esm": "^3.2.25",
+    "gts": "^2.0.2",
+    "prettier": "^2.0.5",
+    "typescript": "^3.8.3"
+  },
+  "engines": {
+    "node": ">12.11.0"
+  },
+  "files": [
+    "./build/src/**/*.ts",
```
Comment (proppy) [inline cli/package.json:41]: should this point to the *.js files?

File: cli/package.json [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    "./build/test/example/**/*.ts"
+  ],
+  "ava": {
+    "files": [
+      "test/**/*",
+      "!test/**/*fixtures.ts"
+    ],
+    "require": [
+      "esm"
```
Comment (proppy) [inline cli/package.json:50]: is that required, I don't see it in https://github.com/avajs/typescript

File: cli/package.json [hunk 11]
```diff
@@ -0,0 +1,58 @@
+{
+  "name": "local-home-testing-cli",
```
Comment (proppy) [inline cli/package.json:2]: @google namespace?

File: cli/src/command-processor.ts [hunk 12]
```diff
@@ -87,6 +87,12 @@ export class CommandProcessor {
       )
       .command('trigger-execute', 'Trigger an Execute command.', yargs => {
         return yargs
+          .option('request_id', {
```
Comment (proppy) [inline cli/src/command-processor.ts:90]: should that go in a separate change?

File: cli/src/platform-worker.ts [hunk 13]
```diff
@@ -59,7 +64,7 @@ export class PlatformWorker {
           // Trigger the identifyHandler with scan results.
           await this.mockLocalHomePlatform.triggerIdentify(
             scanMessage.requestId,
-            scanRemoteInfo.buffer,
+            Buffer.from(scanRemoteInfo.scanData, 'hex'),
```
Comment (proppy) [inline cli/src/platform-worker.ts:67]: what about doing `hex`/`Buffer` related bug fixes in a separate change?

File: stubs/src/radio/radio-controller.ts [hunk 14]
```diff
@@ -142,24 +137,26 @@ export class RadioController {
...
-        socket.send(payload, port, address, error => {
-          if (error) {
-            throw new Error(`Failed to send UDP message: ${error.message}`);
-          }
-          console.log('Sent UDP message: ', payload);
-          if (expectedResponsePackets === 0) {
-            // Resolve early if we aren't expecting any response.
-            resolve(new UdpResponse());
-          }
+        socket.bind(listenPort, () => {
```
Comment (proppy) [inline stubs/src/radio/radio-controller.ts:140]: curious what this fix is about?

File: cli/src/command-processor.ts (hunk 1 above)
Comment (cdaly333) [reply cli/src/command-processor.ts]: Since the TypeScript compiler doesn't resolve the `local-home-testing` local dependency, I believe this import gets resolved at runtime, so I can't target the typescript source.  One option is using `tsconfig` path mapping to hard code the TypeScript source but I'm unsure that would be much better so will leave as-is for now.

File: stubs/src/radio/radio-controller.ts (hunk 2 above)
Comment (cdaly333) [reply stubs/src/radio/radio-controller.ts:10]: Fixed.

File: cli/README.md (hunk 3 above)
Comment (cdaly333) [reply cli/README.md:8]: Will update `README`s in a separate PR.

File: cli/package.json (hunk 4 above)
Comment (cdaly333) [reply cli/package.json:4]: The `local-home-testing-cli` project imports the stubs and true radio functionality from `local-home-testing`, yes.

File: cli/package.json (hunk 5 above)
Comment (cdaly333) [reply cli/package.json]: Removed private.

File: cli/package.json (hunk 6 above)
Comment (cdaly333) [reply cli/package.json:12]: Since we're not using `ts-node` in other projects, I'm inclined to just use `node`.  If I'm not mistaken, the only difference would be the compilation step (and we `include` the build files).

File: cli/package.json (hunk 7 above)
Comment (cdaly333) [reply cli/package.json]: Went through both projects and removed most the `eslint` dependencies.

As a side note, some `eslint` dependencies seem to be a bit outdated in `gts`.  In particular, I had to include the latest version of `@typescript-eslint/eslint-plugin` since it was seemingly required to recognize the Local Home SDK types through triple-slash directives.

File: cli/package.json (hunk 8 above)
Comment (cdaly333) [reply cli/package.json]: Removed.

File: cli/package.json (hunk 9 above)
Comment (cdaly333) [reply cli/package.json]: Fixed

File: cli/package.json (hunk 10 above)
Comment (cdaly333) [reply cli/package.json]: Not required. Removed.

File: cli/package.json (hunk 11 above)
Comment (cdaly333) [reply cli/package.json]: Done.

File: cli/src/command-processor.ts (hunk 12 above)
Comment (cdaly333) [reply cli/src/command-processor.ts:90]: Will include in separate PR.

File: cli/src/platform-worker.ts (hunk 13 above)
Comment (cdaly333) [reply cli/src/platform-worker.ts:67]: Will do.

File: stubs/src/radio/radio-controller.ts (hunk 14 above)
Comment (cdaly333) [reply stubs/src/radio/radio-controller.ts:140]: It's to ensure `socket.send()` is called after `socket.bind()`.  If I'm not mistaken, `socket.send()` will still work without the socket being bound, but it will resort to a default unwanted behavior.  I'll see if I can move this into another PR.

**Comment by cdaly333** [comment]
Removed unrelated changes.  Leaving a few comments unresolved until I open new PRs for the unrelated changed.  All other comments should be addressed.

File: cli/src/command-processor.ts (hunk 12 above)
Comment (cdaly333) [reply cli/src/command-processor.ts:90]: Done

File: stubs/src/radio/radio-controller.ts (hunk 14 above)
Comment (cdaly333) [reply stubs/src/radio/radio-controller.ts]: Done. See #60

File: cli/src/platform-worker.ts (hunk 13 above)
Comment (cdaly333) [reply cli/src/platform-worker.ts]: Done. see #59

File: cli/package.json (hunk 6 above)
Comment (proppy) [reply cli/package.json:12]: This does make the `cli` script dependent on `compile`, should we automatically run it before? or is it documented in the README? (one thing I'm afraid is that since the `cli` script is mainly used for development, one my forget to recompile between each modification).

**Comment by ToniCorinne** [comment]
@cdaly333 got some merge conflicts on here and on #56  and #59 as well - let me know the order in which you want to handle getting those merged.

**Comment by cdaly333** [comment]
Resolved merge conflicts here.  We should merge this one first, followed by #59, and finally #56.  Will need to `rebase` after each merge from here on.
